File: src/domain/quantities.ts
import {
  loadBundledIngredientCatalog,
  type IngredientCatalog,
} from "@/data/ingredientCatalog";
import type { Recipe, RecipeIngredient } from "@/domain/recipes";
import type {
  GeneratedSmoothie,
  GeneratedSmoothieRole,
} from "@/domain/smoothieGenerator";

export const QUANTITY_UNITS = [
  "g",
  "ml",
  "piece",
  "tsp",
  "tbsp",
  "leaf",
  "cube",
] as const;

export type QuantityUnit = (typeof QUANTITY_UNITS)[number];

export interface Quantity {
  readonly amount: number;
  readonly unit: QuantityUnit;
}

export interface QuantifiedIngredient {
  readonly ingredientId: string;
  readonly quantity: Quantity;
}

export interface QuantifiedSmoothie {
  readonly servings: number;
  readonly ingredients: readonly QuantifiedIngredient[];
}

export interface QuantifiedRecipe {
  readonly recipeId: string;
  readonly nameDe: string;
  readonly servings: number;
  readonly required: readonly QuantifiedIngredient[];
  readonly optional: readonly QuantifiedIngredient[];
}

export const MIN_RECIPE_SERVINGS = 1;
export const MAX_RECIPE_SERVINGS = 12;

const UNIT_ALIASES: Readonly<Record<string, QuantityUnit>> = Object.freeze({
  g: "g",
  gram: "g",
  gramm: "g",
  ml: "ml",
  milliliter: "ml",
  piece: "piece",
  pieces: "piece",
  "stück": "piece",
  stueck: "piece",
  tsp: "tsp",
  tl: "tsp",
  "teelöffel": "tsp",
  teeloeffel: "tsp",
  tbsp: "tbsp",
  el: "tbsp",
  "esslöffel": "tbsp",
  essloeffel: "tbsp",
  leaf: "leaf",
  leaves: "leaf",
  blatt: "leaf",
  "blätter": "leaf",
  blaetter: "leaf",
  cube: "cube",
  cubes: "cube",
  "würfel": "cube",
  wuerfel: "cube",
});

export const QUANTITY_UNIT_LABELS_DE: Readonly<Record<QuantityUnit, string>> =
  Object.freeze({
    g: "g",
    ml: "ml",
    piece: "Stück",
    tsp: "TL",
    tbsp: "EL",
    leaf: "Blatt",
    cube: "Würfel",
  });

function normalizeUnitText(value: string): string {
  return value.trim().toLocaleLowerCase("de-DE").replace(/\s+/g, " ");
}

export function normalizeQuantityUnit(value: string | QuantityUnit): QuantityUnit {
  const normalized = normalizeUnitText(value);
  const unit = UNIT_ALIASES[normalized];
  if (!unit) {
    throw new Error(`Unsupported quantity unit: ${value}`);
  }
  return unit;
}

export function createQuantity(
  amount: number,
  unit: string | QuantityUnit,
): Quantity {
  if (!Number.isFinite(amount) || amount <= 0) {
    throw new Error("Quantity amount must be a positive finite number");
  }

  return Object.freeze({
    amount,
    unit: normalizeQuantityUnit(unit),
  });
}

export function quantityLabelDe(quantity: Quantity): string {
  return `${formatQuantityAmount(quantity.amount)} ${QUANTITY_UNIT_LABELS_DE[quantity.unit]}`;
}

export function formatQuantityAmount(amount: number): string {
  if (Number.isInteger(amount)) {
    return String(amount);
  }

  return amount
    .toFixed(amount < 5 ? 2 : 1)
    .replace(/\.0+$/, "")
    .replace(/(\.\d*[1-9])0+$/, "$1")
    .replace(".", ",");
}

function roundAmount(value: number): number {
  if (value >= 20) {
    return Math.round(value);
  }
  if (value >= 5) {
    return Math.round(value * 10) / 10;
  }
  return Math.round(value * 100) / 100;
}

function validateServings(servings: number): void {
  if (
    !Number.isInteger(servings) ||
    servings < MIN_RECIPE_SERVINGS ||
    servings > MAX_RECIPE_SERVINGS
  ) {
    throw new Error(
      `servings must be an integer from ${MIN_RECIPE_SERVINGS} to ${MAX_RECIPE_SERVINGS}`,
    );
  }
}

function roleFraction(
  role: GeneratedSmoothieRole,
  category: string,
): number {
  switch (role) {
    case "main_fruit":
      return 0.5;
    case "liquid":
      return 0.65;
    case "creamy_base":
      return 0.5;
    case "protein":
      return 0.45;
    case "sweetener":
      return 0.2;
    case "extra":
      if (category === "spices") {
        return 0.1;
      }
      if (category === "greens") {
        return 0.45;
      }
      if (category === "seeds") {
        return 0.3;
      }
      return 0.3;
  }
}

function oneServingGeneratedQuantity(
  ingredientId: string,
  role: GeneratedSmoothieRole,
  catalog: IngredientCatalog,
): Quantity {
  const ingredient = catalog.require(ingredientId);
  const range = ingredient.typicalAmount;
  const fraction = roleFraction(role, ingredient.category);
  const amount =
    range.minimum + (range.maximum - range.minimum) * fraction;
  const bounded = Math.min(range.maximum, Math.max(range.minimum, amount));

  return createQuantity(roundAmount(bounded), range.unit);
}

export function quantifyGeneratedSmoothie(
  candidate: GeneratedSmoothie,
  servings = 1,
  catalog: IngredientCatalog = loadBundledIngredientCatalog(),
): QuantifiedSmoothie {
  validateServings(servings);

  const roleByIngredient = new Map(
    candidate.roles.map((assignment) => [
      assignment.ingredientId,
      assignment.role,
    ]),
  );

  const measured = candidate.ingredientIds.map((ingredientId) => {
    const role = roleByIngredient.get(ingredientId);
    if (!role) {
      throw new Error(
        `Generated candidate is missing a role for ingredient: ${ingredientId}`,
      );
    }

    const oneServing = oneServingGeneratedQuantity(
      ingredientId,
      role,
      catalog,
    );

    return Object.freeze({
      ingredientId,
      quantity: createQuantity(
        roundAmount(oneServing.amount * servings),
        oneServing.unit,
      ),
    });
  });

  return Object.freeze({
    servings,
    ingredients: Object.freeze(measured),
  });
}

function quantifyStoredItem(
  item: RecipeIngredient,
  servings: number,
): QuantifiedIngredient {
  return Object.freeze({
    ingredientId: item.ingredientId,
    quantity: createQuantity(
      roundAmount(item.amount * servings),
      item.unit,
    ),
  });
}

export function quantifyStoredRecipe(
  recipe: Recipe,
  servings = 1,
): QuantifiedRecipe {
  validateServings(servings);

  return Object.freeze({
    recipeId: recipe.id,
    nameDe: recipe.nameDe,
    servings,
    required: Object.freeze(
      recipe.required.map((item) => quantifyStoredItem(item, servings)),
    ),
    optional: Object.freeze(
      recipe.optional.map((item) => quantifyStoredItem(item, servings)),
    ),
  });
}

File: smoothie/quantities.py
"""Quantity calculation and serving scaling for generated and stored recipes."""
from __future__ import annotations

from dataclasses import dataclass
from enum import StrEnum

from .generator import GeneratedSmoothie
from .ingredient_catalog import IngredientCatalog
from .ingredients import IngredientCategory
from .recipes import Recipe


class QuantityUnit(StrEnum):
    G = "g"
    ML = "ml"
    PIECE = "piece"
    TSP = "tsp"
    TBSP = "tbsp"
    LEAF = "leaf"
    CUBE = "cube"


_UNIT_ALIASES = {
    "g": QuantityUnit.G,
    "gram": QuantityUnit.G,
    "gramm": QuantityUnit.G,
    "ml": QuantityUnit.ML,
    "milliliter": QuantityUnit.ML,
    "piece": QuantityUnit.PIECE,
    "pieces": QuantityUnit.PIECE,
    "stück": QuantityUnit.PIECE,
    "stueck": QuantityUnit.PIECE,
    "tsp": QuantityUnit.TSP,
    "tl": QuantityUnit.TSP,
    "teelöffel": QuantityUnit.TSP,
    "teeloeffel": QuantityUnit.TSP,
    "tbsp": QuantityUnit.TBSP,
    "el": QuantityUnit.TBSP,
    "esslöffel": QuantityUnit.TBSP,
    "essloeffel": QuantityUnit.TBSP,
    "leaf": QuantityUnit.LEAF,
    "leaves": QuantityUnit.LEAF,
    "blatt": QuantityUnit.LEAF,
    "blätter": QuantityUnit.LEAF,
    "blaetter": QuantityUnit.LEAF,
    "cube": QuantityUnit.CUBE,
    "cubes": QuantityUnit.CUBE,
    "würfel": QuantityUnit.CUBE,
    "wuerfel": QuantityUnit.CUBE,
}

_UNIT_LABELS_DE = {
    QuantityUnit.G: "g",
    QuantityUnit.ML: "ml",
    QuantityUnit.PIECE: "Stück",
    QuantityUnit.TSP: "TL",
    QuantityUnit.TBSP: "EL",
    QuantityUnit.LEAF: "Blatt",
    QuantityUnit.CUBE: "Würfel",
}


@dataclass(frozen=True)
class Quantity:
    amount: float
    unit: QuantityUnit

    def __post_init__(self) -> None:
        if self.amount <= 0:
            raise ValueError("Quantity amount must be positive")

    @classmethod
    def from_raw(cls, amount: float, unit: str | QuantityUnit) -> "Quantity":
        if isinstance(unit, QuantityUnit):
            return cls(amount, unit)
        normalized = " ".join(unit.strip().casefold().split())
        try:
            canonical = _UNIT_ALIASES[normalized]
        except KeyError as exc:
            raise ValueError(f"Unsupported quantity unit: {unit!r}") from exc
        return cls(amount, canonical)

    @property
    def label_de(self) -> str:
        return _UNIT_LABELS_DE[self.unit]


@dataclass(frozen=True)
class QuantifiedIngredient:
    ingredient_id: str
    quantity: Quantity


@dataclass(frozen=True)
class QuantifiedSmoothie:
    servings: int
    ingredients: tuple[QuantifiedIngredient, ...]


@dataclass(frozen=True)
class QuantifiedRecipe:
    recipe_id: str
    name_de: str
    servings: int
    required: tuple[QuantifiedIngredient, ...]
    optional: tuple[QuantifiedIngredient, ...]


class QuantityCalculator:
    """Convert recipe structures into practical, serving-scaled quantities."""

    def __init__(self, catalog: IngredientCatalog) -> None:
        self.catalog = catalog

    def for_generated(self, candidate: GeneratedSmoothie, servings: int = 1) -> QuantifiedSmoothie:
        self._validate_servings(servings)
        roles = dict(candidate.roles)
        measured: list[QuantifiedIngredient] = []
        for ingredient_id in candidate.ingredient_ids:
            ingredient = self.catalog.require(ingredient_id)
            role = roles[ingredient_id]
            one_serving = self._one_serving_amount(ingredient_id, role)
            quantity = Quantity.from_raw(
                self._round_amount(one_serving * servings),
                ingredient.typical_amount.unit,
            )
            measured.append(QuantifiedIngredient(ingredient_id, quantity))
        return QuantifiedSmoothie(servings, tuple(measured))

    def for_stored(self, recipe: Recipe, servings: int = 1) -> QuantifiedRecipe:
        self._validate_servings(servings)

        def scale(item) -> QuantifiedIngredient:
            return QuantifiedIngredient(
                item.ingredient_id,
                Quantity.from_raw(self._round_amount(item.amount * servings), item.unit),
            )

        return QuantifiedRecipe(
            recipe_id=recipe.id,
            name_de=recipe.name_de,
            servings=servings,
            required=tuple(scale(item) for item in recipe.required),
            optional=tuple(scale(item) for item in recipe.optional),
        )

    def _one_serving_amount(self, ingredient_id: str, role: str) -> float:
        ingredient = self.catalog.require(ingredient_id)
        amount = ingredient.typical_amount
        fraction = {
            "main_fruit": 0.50,
            "liquid": 0.65,
            "creamy_base": 0.50,
            "protein": 0.45,
            "sweetener": 0.20,
        }.get(role)
        if fraction is None:
            if ingredient.category == IngredientCategory.SPICES:
                fraction = 0.10
            elif ingredient.category == IngredientCategory.GREENS:
                fraction = 0.45
            elif ingredient.category == IngredientCategory.SEEDS:
                fraction = 0.30
            else:
                fraction = 0.30
        value = amount.minimum + (amount.maximum - amount.minimum) * fraction
        return min(amount.maximum, max(amount.minimum, value))

    @staticmethod
    def _round_amount(value: float) -> float:
        if value >= 20:
            return float(round(value))
        if value >= 5:
            return round(value, 1)
        return round(value, 2)

    @staticmethod
    def _validate_servings(servings: int) -> None:
        if not isinstance(servings, int) or isinstance(servings, bool) or servings < 1:
            raise ValueError("servings must be a positive integer")
